src/api/Controllers/CoursesController.cs

```diff
@@ -256,30 +257,31 @@ private void AssociateWithLocations(ref IList<Course> courses)
                 .Concat(courses.Select(x=>x.ProviderLocation?.Address))
                 .Where(x =>  !string.IsNullOrWhiteSpace(x))
                 .Distinct()
-                .ToDictionary(x => x, x => new Location { Address = x });
+                .ToDictionary(x => x, x => {
+                    var location = _context.Locations.FirstOrDefault(l => l.Address == x);
```

Comment:
n+1 fetch, very inefficient. cache with `var allExistingLocations = _context.Locations.ToList()`. Then make it a lookup by saying `ToLookup(x=>x.Address)` and to be resilient to redundant data, only take the first Location object, `ToDictionary(x => x.Key, x.First())`.

reject